File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
class Solution {
        public int largestRectangleArea(int[] heights) {
        int len = heights.length;
        int[] left = new int[len];
        int[] right = new int[len];
        Stack<Integer> stack = new Stack<>();
        for(int i=0;i<len;i++){
            if(stack.isEmpty()){
                stack.push(i);
                left[i] = 0;
            }else{
                while(!stack.isEmpty() && heights[stack.peek()] >= heights[i])
                    stack.pop();
                left[i] = stack.isEmpty()?0:stack.peek()+1;
                stack.push(i);
            }
        }
        stack = new Stack<>();
        for(int i=len-1;i>=0;i--){
            if(stack.isEmpty()){
                stack.push(i);
                right[i] = len-1;
            }else{
                while(!stack.isEmpty() && heights[stack.peek()] >= heights[i])
                    stack.pop();
                right[i] = stack.isEmpty()?len-1:stack.peek()-1;
                stack.push(i);
            }
        }
        int maxArea = Integer.MIN_VALUE;
        int[] area = new int[len];
        for(int i=0;i<len;i++){
            area[i] = (right[i] - left[i] + 1) * heights[i];
            maxArea = Math.max(maxArea, area[i]);
        }
        return maxArea;
    }
}

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []  # (start_index, height)
        max_area = 0

        for i, h in enumerate(heights):
            start = i

            while stack and stack[-1][1] > h:
                index, height = stack.pop()

                width = i - index
                area = height * width
                max_area = max(max_area, area)

                start = index

            stack.append((start, h))

        n = len(heights)

        while stack:
            index, height = stack.pop()

            width = n - index
            area = height * width
            max_area = max(max_area, area)

        return max_area